java/runtime/src/main/java/io/ray/runtime/metric/MetricRegistry.java

```diff
@@ -0,0 +1,136 @@
+    startup(MetricConfig.DEFAULT_CONFIG);
+  }
+
+  public void startup(MetricConfig metricConfig) {
+    synchronized (this) {
+      if (!isRunning) {
+        this.metricConfig = metricConfig;
+        scheduledExecutorService = new ScheduledThreadPoolExecutor(metricConfig.threadPoolSize(),
+          new ThreadFactoryBuilder().setNameFormat("metric-registry-%d").build());
+        scheduledExecutorService.scheduleAtFixedRate(this::update, metricConfig.timeIntervalMs(),
+          metricConfig.timeIntervalMs(), TimeUnit.MILLISECONDS);
+        isRunning = true;
+        LOG.info("Finished startup metric registry, metricConfig is {}.", metricConfig);
+      }
+    }
+  }
+
+  public void shutdown() {
+    synchronized (this) {
+      if (isRunning && scheduledExecutorService != null) {
+        try {
+          scheduledExecutorService.shutdownNow();
+          if (!scheduledExecutorService
+            .awaitTermination(metricConfig.shutdownWaitTimeMs(), TimeUnit.MILLISECONDS)) {
+            LOG.warn("Metric registry did not shut down in {}ms time, so try to shut down again.",
+              metricConfig.shutdownWaitTimeMs());
+            scheduledExecutorService.shutdownNow();
+          }
+        } catch (InterruptedException e) {
+          LOG.warn("Interrupted when shutting down metric registry, so try to shut down again.",
+            e.getMessage(), e);
+          scheduledExecutorService.shutdownNow();
+        }
+        if (scheduledExecutorService.isShutdown()) {
+          isRunning = false;
+          scheduledExecutorService = null;
+          LOG.info("Metric registry has been shut down.");
+        } else {
+          LOG.warn("Failed to shut down metric registry");
+        }
+      }
+    }
+  }
+
+  public Metric register(Metric metric) {
+    synchronized (this) {
+      if (!isRunning) {
+        LOG.warn("Failed to register a metric, because the metric registry is not running.");
+        return null;
+      }
+      try {
+        MetricId id = genMetricIdByMetric(metric);
+        Metric ori = registeredMetrics.putIfAbsent(id, metric);
+        if (ori == null) {
+          return metric;
+        } else {
+          LOG.info("Metric {} has already registered, so use the previous one.", id);
+          return ori;
+        }
+      } catch (Exception e) {
+        LOG.warn("Failed to register a metric: ", e.getMessage(), e);
+        return null;
+      }
+    }
+  }
+
+  public void unregister(Metric metric) {
+    synchronized (this) {
+      if (!isRunning) {
+        LOG.warn("Failed to unregister a metric, because the metric registry is not running.");
+      }
+      try {
+        MetricId id = genMetricIdByMetric(metric);
+        registeredMetrics.remove(id);
+      } catch (Exception e) {
+        LOG.warn("Failed to unregister a metric: ", e.getMessage(), e);
+      }
+    }
+  }
+
+  private void update() {
+    registeredMetrics.forEach((id, metric) -> {
+      metric.record();
```

Comment:
If we update metric value between record and reset, what's happened?

Reply:
The update value will lost. Its' a problem. I'll fix it.

Reply:
Fixed.